### DISCUSSION TITLE
q-uploader - multi-upload, custom fields for each file

### DISCUSSION THREAD

**Discussion by bitmandk** [quasarframework/quasar#15061] [General - Components / Directives / etc] [answered]
How can I access additional data for each file uploaded?

For each file I have a DocumentType and a Remark, that I want to submit to the endpoint either together with each fileupload or after each fileupload has finished succesfully.

I have the below uploader (CodePen here: https://codepen.io/bobittjek/pen/poKqwRr)

![image](https://user-images.githubusercontent.com/7941347/205693259-cbd6e086-4f85-4b91-a750-884d07440d5e.png)

According to the documentation: "These two props (headers and form-fields) can be used as a function too ((files) => Array), allowing you to dynamically set them based on the files that are to be uploaded."

But the above dosn't make sense to me -  how do I set them dynamically for each uploaded file (remark and documenttype)?

**Comment by bitmandk** [comment]
I have an updated CodePen: https://codepen.io/bobittjek/pen/KKebLpj (documentType removed for readability)

In this pen v-model for the remark textbox now points to an array that is created when files are added - so I can keep track of the remarks added to each file:

**Template additions:**
```
<q-uploader .... @added="onFilesAdded">
.
.
<template v-for="(file, fileIndex) in scope.files">
.
.
<q-input v-model="remarkModel[fileIndex]"></q-input>
```
.
.
.
**Script additions:**
```
const data = {
 remarkModel: []
}
.
.
 onFilesAdded(files) {
                for (var i = 0; i < files.length; i++) {
                    this.remarkModel.push('')
                }
            }
```

Only problem left, how can I attach each v-model (remark) to the form-fields that is submitted to the endpoint?

**Comment by metalsadman** [comment, marked as answer]
@bitmandk this should get you on https://codepen.io/metalsadman/pen/oNyJGve
upload and check the payload, see your extra form fields passed.

**Reply by bitmandk** [reply]
@metalsadman thanks for the "pen" - a bit more elegant than what I was trying to do above.

Your CodePen seems to work fine - I have implemented my upload in a VUE component, so I've tried to convert your pen to be used in a compoment. It runs without problems, but "remarks" field and "documentType" "forgets" my input when loosing focus (see implementation below). 

When I check VueDevTools both "remarks" and "documentType" are still NULL. If I manually edit the "remarks" property using DevTools - the remark does now show up in the textfield, but I can see the remark in my template debug code, really strange:

I'm using options API, so I've tried to convert it into that.
 
![image](https://user-images.githubusercontent.com/7941347/205926423-2a175115-461b-4208-a9ea-c93a74b56544.png)

**Component implementation**

```
<template>
    <div>
        <q-uploader style="width: auto;" v-bind:url="uploadEndpoint()"
                    label="Tilføj dokumenter"
                    multiple
                    :form-fields="formFields"
                    @added="onAdd"
                    @removed="onRemove">
            <template v-slot:list="scope">
                <q-list separator>
                    <template v-for="file in scope.files" :key="file.__key">
                        <q-item>
                            <q-item-section>
                                <q-item-label class="full-width ellipsis">
                                    {{ file.name }}
                                </q-item-label>

                                <q-item-label caption>
                                    Status: {{ file.__status }}
                                </q-item-label>

                                <q-item-label caption>
                                    {{ file.__sizeLabel }} / {{ file.__progressLabel }}
                                </q-item-label>
                            </q-item-section>
                            <q-item-section v-if="file.__img"
                                            thumbnail
                                            class="gt-xs">
                                <img :src="file.__img.src">
                            </q-item-section>
                            <q-item-section top side>
                                <q-btn class="gt-xs"
                                       size="12px"
                                       flat
                                       dense
                                       round
                                       icon="delete"
                                       @click="scope.removeFile(file)">
                                    <q-tooltip>Fjern dokument fra upload kø
                                </q-btn>
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>
                                <q-select v-model="allFiles.find(v => v.__key === file.__key).docType"
                                          square outlined dense
                                          label="Dokumenttype"
                                          :options="documentTypes"
                                          option-value="documentTypeId"
                                          option-label="name"
                                          emit-value>
                                </q-select>
                            </q-item-section>
                            <q-item-section>
                                <q-input v-model="allFiles.find(v => v.__key === file.__key).remarks"
                                         clearable square outlined dense label="Bemærkning">
                                </q-input>
                            </q-item-section>
                        </q-item>
                    </template>
                </q-list>
            </template>
        </q-uploader>
    </div>
</template>
<script>
    var apiBase = internalApiV1BaseURL
    export default {
        props: {
            caseGuid: String,
            documentTypes: Object
        },
        data() {
            return {
                allFiles: []
                //     setup
            };
        },
        methods: {
            uploadEndpoint() {
                return apiBase + "/cases/" + this.caseGuid + "/documents/upload"
            },
            onAdd(files) {
                console.log(files)
                this.allFiles.push(...(files.map(v => ({ ...v, docType: null, remarks: null }))))
            },
            onRemove(files) {
                this.allFiles = this.allFiles.filter(v => {
                    return files.some(f => v.__key !== f.__key)
                })
            },
            formFields(files) {
                // get the file from list of files
                const dFile = this.allFiles.value.find(v => v.__key === files[0].__key)
                // assign the extra props to form data
                return [{ name: 'DocumentType', value: dFile.docType }, { name: 'Remarks', value: dFile.remarks }]
            }
        }
    }
</script>
```

**Reply by metalsadman** [reply]
here `const dFile = this.allFiles.value.find(v => v.__key === files[0].__key)` since you're using options api, remove `.value`
your `documentTypes` is object form, if it's an object and array inside, then use that as your `:options` prop. you're close to what i've written in the codepen, just have to be mindful of some nuances when converting, i wonder why you're not using composition api tho.

**Reply by bitmandk** [reply]
I found this earlier, but haven't updated the code above : `const dFile = this.allFiles.value.find(v => v.__key === files[0].__key)`but removing .value didn't solve it. Have never used composition.

I'll try to make a CodePen that reflects the component.

**Comment by bitmandk** [comment]
Not Quasar related, but really strange, it looks like the v-model "allFiles.find" dosn't find anything, so I changed it into this instead, which also dosn't find any match:

```
 <q-input :v-model="modelRemark(file.__key)"
                                         clearable square outlined dense label="Bemærkning">
                                </q-input>
.
.
 modelRemark(fileKey) {
                console.log("KEY: " + fileKey)
                var result = this.allFiles.find(v => v.__key === fileKey)
                console.log("Result KEY: " + result.__key)
                console.log("Remarks: " + result.remarks)
                return result.remarks
            }
```
Have the following output in DevTools and console:

DevTools - Allfiles, contains data:
![image](https://user-images.githubusercontent.com/7941347/205953348-14c154c5-05d2-4228-8970-aab6985363b8.png)

Console - debug output from modelRemark function:
![image](https://user-images.githubusercontent.com/7941347/205953562-7ebecf87-7da5-4bc2-907d-b851385686ef.png)

If I edit the allFiles entry manually using DevTools, the change shows up in debug and console but not in the remarks textinput:

![image](https://user-images.githubusercontent.com/7941347/205952975-7ad37363-c07e-4b70-9b29-23c58340331e.png)

**Reply by metalsadman** [reply]
`:v-model` is a syntax error. are you using quasar v1?

**Reply by bitmandk** [reply]
No, I'm using: 2.8.3

```
 <q-input v-model="modelRemark(file.__key)" clearable square outlined dense label="Bemærkning">
</q-input>
```
When using the above, nothing is shown in the console:
![image](https://user-images.githubusercontent.com/7941347/205992744-5fa782d1-f748-40fe-9572-8b950134621f.png)

But if I use `:-vmodel`:

```
<q-input :v-model="modelRemark(file.__key)" clearable square outlined dense label="Bemærkning">
</q-input>

```
Then I get output in the console (from the modelRemark method):

![image](https://user-images.githubusercontent.com/7941347/205993086-a14d6940-0922-4859-a186-fdd5216d5763.png)

**Comment by bitmandk** [comment]
@metalsadman

Sorry for the mixeup code-versions, here's an update CodePen, using Options API and implemented as UMD:

https://codepen.io/bobittjek/pen/LYrqRXZ

**Comment by pdanpdan** [comment]
You made a total mess on what's in there :)))

1. (not in codepen - just to understand) You cannot use the the file in scope to directly modify it (the semantic is of a readonly variable)
2. You had functions that returned a value and used them as v-model - what would you expect to be modified (it's like you would say `<input v-model="'constant text'" />`)
3. don't calculate lots of thing in loops if they can be precalculated
4. you didn't have a allFiles initialized in data, so not reactive (and the push in onAdd should error)
5. ...


https://codepen.io/pdanpdan/pen/mdKvOMj?editors=1010

**Reply by bitmandk** [reply]
Both your CodePen and the one from metalsadman works fine. But when I try to implement it 1:1 in my .VUE component (SFC) it doesn't work.

I'll try to go through the implemented code line by line tomorrow to see what I'm missing since it dosn't work, when used as a component.

**Component code:**
```
<template>
    <div>
        <q-uploader style="width: auto;" url="uploadEndpoint"
                    label="Tilføj dokumenter"
                    multiple
                    :form-fields="formFields"
                    @added="onAdd"
                    @removed="onRemove">
            <template v-slot:list="scope">
                <q-list separator>
                    <template v-for="file in scope.files" :key="file.__key">
                        <q-item>
                            <q-item-section>
                                <q-item-label class="full-width ellipsis">
                                    {{ file.name }}
                                </q-item-label>

                                <q-item-label caption>
                                    Status: {{ file.__status }}
                                </q-item-label>

                                <q-item-label caption>
                                    {{ file.__sizeLabel }} / {{ file.__progressLabel }}
                                </q-item-label>
                            </q-item-section>
                            <q-item-section v-if="file.__img"
                                            thumbnail
                                            class="gt-xs">
                                <img :src="file.__img.src">
                            </q-item-section>
                            <q-item-section top side>
                                <q-btn class="gt-xs"
                                       size="12px"
                                       flat
                                       dense
                                       round
                                       icon="delete"
                                       @click="scope.removeFile(file)">
                                    <q-tooltip>Fjern dokument fra upload kø
                                </q-btn>
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>
                                {{allFiles}}<br />
                                Remark:{{allFiles[0].remarks}}
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>

                                <q-select v-model="getFile(file.__key).docType"
                                       square outlined dense
                                       label="Dokumenttype"
                                       :options="documentTypes"
                                       option-value="documentTypeId"
                                       option-label="name">
                             </q-select>
                            </q-item-section>
                            <q-item-section>
                                <q-input v-model="getFile(file.__key).remark"
                                         clearable square outlined dense label="Bemærkning">
                                </q-input>
                            </q-item-section>
                        </q-item>
                    </template>
                </q-list>
            </template>
        </q-uploader>
    </div>
</template>
<script>
    var apiBase = internalApiV1BaseURL
    export default {
        props: {
            caseGuid: String,
            documentTypes: Array
        },
        data() {
            return {
                allFiles: []
            };
        },
        computed: {
            uploadEndpoint() {
                return apiBase + '/cases/${encodeURIComponent(this.caseGuid)}/documents/upload'
            }
        },
        methods: {
            getFile(key) {
                return this.allFiles.find(f => f.__key === key)
            },
            onAdd(files) { //preselect document with documentTypeId == 1
                const docType = null //documentTypes.find(d => d.value === 1)
                this.allFiles = this.allFiles.concat(
                    files.map(v => ({
                        ...v,
                        docType,
                        remarks: null
                    }))
                )
            },
            onRemove(files) {
                const keys = files.map(f => f.__key)
                this.allFiles = this.allFiles.filter(v => keys.indexof(v.__key) === -1)
            },
            formFields(files) {
                // get the file from list of files
                const key = files[0].__key
                const file = this.allFiles.find(v => v.__key === key)
                // assign the extra props to form data
                return [{ name: 'DocumentType', value: file.docType.value }, { name: 'Remarks', value: file.remarks }]
            }
        }
    }
</script>
```

**Reply by metalsadman** [reply]
I don't see why it would only work in codepen and not on your code, you haven't stated any error log other than simply stating "it doesn't work". I think we've provided enough insight and will mark this as answered. Goodluck.

**Reply by bitmandk** [reply]
I appreciate your help, but there are no errors to post. As I wrote earlier when I enter a value in the remark text input, it disapears when loosing focus for some strange reason. So only solution is to try to implement both of your solutions and see why the text input fails to "remeber" the entered value . So nothing more you can do and thanks for helping in the right direction.

**Comment by bitmandk** [comment]
Just to wrap this up:

Both solutions suggested by @metalsadman and @pdanpdan works perfectly in CodePen and when implemented in a single html file, thank you!

My problem was (still is), that if I take the suggested code from CodePen and more or less implement it 1:1 in a .VUE component then the suggested solutions does not work - Everything works, except - the remarks and documenttypes are not "remembered" when focus is lot from the input fiels. No errormessages or anything is shown.

The problems are probably related to .VUE components and should probably be posted in a VUE forum, but if anybody is interested, feel free to take a look at it.

- **Upload test** - single page -this works 100% as proposed:
Link removed

- **Upload test as component** - suggested solution implemented as .VUE component and loaded with vue3-sfc-loader.js. Only minimal changes to code in order to use it as a component. Upload works, errors in console "[Vue warn]: Missing required prop: 'modelValue'" Values entered in input fields are not remembered when input field looses focus.
Link removed
Really strange.....

**Comment by pdanpdan** [comment]
Post the component sfc

**Reply by bitmandk** [reply]
@pdanpdan 
If I comment-out the SELECT component, console warning disapears - but still text is lost when loosing focus.
Looks like the v-model 2-waybinding does not work when used in a component.

I just noticed this in your CodePen remarks/remark:

![image](https://user-images.githubusercontent.com/7941347/206193475-48b3323f-ef4d-4d0d-be28-9558250c9a39.png)

But it dosn't affect the v-model as it's being created when referenced.


Component code:
```
<script>
    export default {
            data() {
            return {
                allFiles: [],
                documentTypes: [
                    { value: 1, label: 'Type 1' },
                    { value: 2, label: 'Type 2' }
                ],
                caseGuid: '686296F3-EEC5-4331-8A32-C885815297F6'
            };
            },
            computed: {
                uploadEndpoint() {
                    return `https://localhost/cases/${encodeURIComponent(this.caseGuid)}/documents/upload`
                }
            },
            methods: {
                getFile(key) {
                    return this.allFiles.find(f => f.__key === key)
                },

                onAdd(files) { //preselect document with documentTypeId == 1
                    const docType = this.documentTypes.find(d => d.value === 1)
                    this.allFiles = this.allFiles.concat(
                        files.map(v => ({
                            ...v,
                            docType,
                            remarks: null
                        }))
                    )
                },

                onRemove(files) {
                    const keys = files.map(f => f.__key)
                    this.allFiles = this.allFiles.filter(v => keys.indexof(v.__key) === -1)
                },

                formFields(files) {
                    // get the file from list of files
                    const key = files[0].__key
                    const file = this.allFiles.find(v => v.__key === key)
                    // assign the extra props to form data
                    return [{ name: 'DocumentType', value: file.docType.value }, { name: 'Remarks', value: file.remarks }]
                }
            }
        }
</script>
<template>
    <div>
        <q-uploader style="width: auto; max-width: 600px"
                    :url="uploadEndpoint"
                    label="Tilføj dokumenter"
                    multiple
                    :form-fields="formFields"
                    @added="onAdd"
                    @removed="onRemove">
            <template v-slot:list="scope">
                <q-list separator>
                    <template v-for="file in scope.files" :key="file.__key">
                        <q-item>
                            <q-item-section>
                                <q-item-label class="full-width ellipsis">{{ file.name }}</q-item-label>
                                <q-item-label caption>Status: {{ file.__status }}</q-item-label>
                                <q-item-label caption>{{ file.__sizeLabel }} / {{ file.__progressLabel }}</q-item-label>
                            </q-item-section>
                            <q-item-section v-if="file.__img && $q.screen.gt.xs" thumbnail>
                                <img :src="file.__img.src">
                            </q-item-section>
                            <q-item-section top side>
                                <q-btn class="gt-xs"
                                       size="12px"
                                       flat
                                       dense
                                       round
                                       icon="delete"
                                       @click="scope.removeFile(file)">
                                    <q-tooltip>Fjern dokument fra upload kø
                                </q-btn>
                            </q-item-section>
                        </q-item>
                        <q-item>
                            <q-item-section>
                                <q-select v-model="getFile(file.__key).docType"
                                          square
                                          outlined
                                          dense
                                          label="Dokumenttype"
                                          :options="documentTypes"></q-select>
                            </q-item-section>
                            <q-item-section>
                                <q-input v-model="getFile(file.__key).remark"
                                         clearable
                                         square
                                         outlined
                                         dense
                                         label="Remark"></q-input>
                            </q-item-section>
                        </q-item>
                    </template>
                </q-list>
            </template>
        </q-uploader>
    </div>
</template>
```

**Comment by metalsadman** [comment]
@bitmandk gotta watch out for typos and take heed of the logs.
these are your errors:
![image](https://user-images.githubusercontent.com/31038838/206197585-4ed48446-8869-4a96-a530-f941febcf326.png)
it's no wonder why your code weren't indented properly :/

and another
![image](https://user-images.githubusercontent.com/31038838/206197791-64109ed9-e985-4e1d-bafa-bc616393e6eb.png)
should be `remarks`

see in action 
![uploader conundrum](https://user-images.githubusercontent.com/31038838/206198443-98e59893-eaa7-4506-90c1-068cc8a0a289.gif)

code here https://stackblitz.com/edit/quasarframework-avafjp?file=src%2Fpages%2FIndexPage.vue

**Reply by bitmandk** [reply]
@metalsadman Thanks for you continued effort on this. The remarks/remark typo was found earlier and fixed, but that didn't solve my problem.

I see that you implementation also works when in a component. But I still have the same problem - either because I use vue-sfc loader to load the component or because of something I have overlooked in the page that loads the component.

I have pasted your component code 1:1 to this file:
Link removed

Same issue. But the problem must be down to something else in the page that loads the component - not Quasar or VUE related.

thanks again.

**Reply by metalsadman** [reply]
create a repository, I find it weird that it wouldn't work in your case being it working in the samples we provided. I don't see it as a bug in Vue either, other than maybe within your code/setup is where something's wrong.

**Reply by metalsadman** [reply]
tested it locally in `dev` and `build` as well and ran without problems.

**Reply by bitmandk** [reply]
I'm currently creating a repository, So I'll see how that goes.

**Reply by bitmandk** [reply]
@metalsadman 
Here's the repository:

https://github.com/bitmandk/QuasarUpload

**Comment by bitmandk** [comment]
This is driving me crazy. All above solutuions by @metalsadman and @pdanpdan works. But when used in a component loaded with VUE-sfc-loader it doesn't work (no erros). I have recreated the problem is this repository: 
https://github.com/bitmandk/QuasarUpload

Really strange!

**Comment by bitmandk** [comment]
Adding to this weird problem.

If I replace this:  
`<q-input  v-model="getFile(file.__key).remarks" label="Remark"></q-input>`

With this: 
`<input type="text" v-model="getFile(file.__key).remarks">`

Then the value is not lost, but it's still not stored in the allFiles data:

![image](https://user-images.githubusercontent.com/7941347/206696377-cc94ae5c-beab-47ef-a721-1958196d6021.png)
